# JupyterLab Connection

In [1]:
%%!
pip install --trusted-host pypi.org ipython-sql

['Requirement already satisfied: ipython-sql in c:\\users\\obo\\anaconda3\\lib\\site-packages (0.4.0)',
 'Requirement already satisfied: six in c:\\users\\obo\\anaconda3\\lib\\site-packages (from ipython-sql) (1.15.0)',
 'Requirement already satisfied: sqlparse in c:\\users\\obo\\anaconda3\\lib\\site-packages (from ipython-sql) (0.4.1)',
 'Requirement already satisfied: prettytable<1 in c:\\users\\obo\\anaconda3\\lib\\site-packages (from ipython-sql) (0.7.2)',
 'Requirement already satisfied: ipython-genutils>=0.1.0 in c:\\users\\obo\\anaconda3\\lib\\site-packages (from ipython-sql) (0.2.0)',
 'Requirement already satisfied: sqlalchemy>=0.6.7 in c:\\users\\obo\\anaconda3\\lib\\site-packages (from ipython-sql) (1.4.13)',
 'Requirement already satisfied: ipython>=1.0 in c:\\users\\obo\\anaconda3\\lib\\site-packages (from ipython-sql) (7.19.0)',
 'Requirement already satisfied: decorator in c:\\users\\obo\\anaconda3\\lib\\site-packages (from ipython>=1.0->ipython-sql) (4.4.2)',
 'Requirem

In [2]:
%load_ext sql

In [4]:
%sql sqlite:///etrade.db

# DATA PREPARATION

I simulated an e-trade  relational database which have 17 tables . Some of them will be created bu JupyterLab  and some of them will be formed in SQLlite.I found table's data from internet.

**Important notes: The names,surnames and all personel data are faked.They were constituted by scprits and i am gonna share shopping scripts(for MSSQL) which chose a random user and appoint the items for shopper in a random time.**

# CRETAION OF TABLES

In [ ]:
%%sql

#CREATE USER TABLE

CREATE TABLE user (
ID INT IDENTITY(1,1),
USERNAME_ VARCHAR(50),
PASSWORD_ VARCHAR(50),
NAMESURNAME VARCHAR(100),
EMAIL VARCHAR(100),
GENDER VARCHAR(1),
CREATEDDATE DATETIME,
BIRTHDATE DATE,
TELNR1 VARCHAR(15),
TELNR2 VARCHAR(15),
CONSTRAINT [PK_USER_] PRIMARY KEY  (ID ASC))


In [23]:
%%sql

#CREATE COUNTRY,CITY,TOWN,DISTRICT
CREATE TABLE COUNTRY(
ID INT IDENTITY(1,1),
COUNTRY VARCHAR(100),
CONSTRAINT [PK_COUNTRY] PRIMARY KEY  (ID ASC));

CREATE TABLE TOWN(
ID INT IDENTITY(1,1),
CITYID INT,
TOWN VARCHAR(100),
CONSTRAINT [PK_TOWN] PRIMARY KEY  (ID ASC));

CREATE TABLE DISTRICT(
ID INT IDENTITY(1,1),
TOWNID INT,
DISTRICT VARCHAR(100),
CONSTRAINT [PK_DISTRICT] PRIMARY KEY  (ID ASC));

CREATE TABLE CITY(
ID INT IDENTITY(1,1),
CITY VARCHAR(100),
CONSTRAINT [PK_CITY] PRIMARY KEY  (ID ASC));


 * sqlite:///etrade.db
Done.
Done.
Done.
Done.


ResourceClosedError: This result object does not return rows. It has been closed automatically.

I created a few table from here and the others ones are formed on the SQLite. Let's review tables

## REVIEW ALL TABLES

In [188]:
%%sql
SELECT * FROM user LIMIT 3

 * sqlite:///etrade.db
Done.


[(1, 'N_OZSIMITCI', 'doropipa', 'Nazlıcan ÖZSİMİTÇİ', 'N_OZSIMITCI@sqlegitim.com', 'K', '2018-07-25 01:35:35.000', '1951-12-29', '(555)3852463', '(532)5898448'),
 (2, 'A_UNLUMAMULERI', 'boerke80', 'Arya UNLUMAMULERİ', 'A_UNLUMAMULERI@sqlegitim.com', 'K', '2018-07-26 01:35:35.000', '1957-03-25', '(537)5169149', '(544)8434440'),
 (3, 'D_GURBETOGLU', 'theebryantbitch', 'Döne GURBETOĞLU', 'D_GURBETOGLU@sqlegitim.com', 'K', '2018-07-27 01:35:35.000', '1986-08-10', '(534)9106933', '(505)1625581')]

This table has knowledge about users who use e-trade website.As i mentioned above, the whole personel data is simulated and they are not real people.

In [38]:
%%sql
SELECT * FROM ADDRESS LIMIT 3

 * sqlite:///etrade.db
Done.


[(1, 1, 0, 1, 41, 252, 100231, 100231, '?HSAN?YE MERKEZ MAH.4375.. SOKAK NO:607 100231 G�LC�K/KOCAEL?'),
 (2, 1, 1, 1, 71, 363, 116027, 116027, 'FAT?H MAH.2757. SOKAK NO:856 116027 KIRIKKALE MERKEZ/KIRIKKALE'),
 (3, 1, 2, 1, 59, 148, 110896, 110896, 'GAZ? MUSTAFA KEMALPA?A MAH.KARACAN SOKAK NO:160 110896 �ERKEZK�Y/TEK?RDA?')]

In [5]:
%%sql

SELECT * FROM COUNTRY LIMIT 3

 * sqlite:///etrade.db
Done.


[(1, 'TURKEY')]

This table has knowledge about country which includes just Turkey because the whole address are in Turkey. 

In [5]:
%%sql
SELECT * FROM CITY LIMIT 3

 * sqlite:///etrade.db
Done.


[(1, 'ADANA'), (2, 'ADIYAMAN'), (3, 'AFYONKARAHİSAR')]

This table has knowledge about cities which are in  Turkey.

In [39]:
%%sql
SELECT * FROM TOWN LIMIT 3

 * sqlite:///etrade.db
Done.


[(1, 37, 'ABANA'), (2, 20, 'ACIPAYAM'), (3, 34, 'ADALAR')]

This table has knowledge about towns  which belongs Turkey's cities.

In [221]:
%%sql
SELECT * FROM DISTRICT LIMIT 3

 * sqlite:///etrade.db
Done.


[(71204, 20, 'ACIPINAR-BAHÇELİEVLER MAH.'),
 (71205, 20, 'ACIPINAR-CERİT MAH.'),
 (71206, 20, 'ACIPINAR-FATİH MAH.')]

This table has knowledge about districts  which belongs Turkey's cities' towns.

In [12]:
%%sql
SELECT * FROM DISTRICT_STREET LIMIT 3

 * sqlite:///etrade.db
Done.


[(1, 1, 473, 3195, 78655, 'ADANA', 'POZANTI', 'AKÇATEKİR', 'AKÇATEKİR-HACILAR MAH.', '607. SOKAK'),
 (2, 1, 473, 3195, 78655, 'ADANA', 'POZANTI', 'AKÇATEKİR', 'AKÇATEKİR-HACILAR MAH.', '630. SOKAK'),
 (3, 1, 473, 3195, 78655, 'ADANA', 'POZANTI', 'AKÇATEKİR', 'AKÇATEKİR-HACILAR MAH.', '602. SOKAK')]

This table has knowledge about districts  which belongs Turkey's cities' towns' districts' streetname.

In [14]:
%%sql
SELECT * FROM ITEM LIMIT 3

 * sqlite:///etrade.db
Done.


[(1, 25655, 'Reis Gönen Baldo Pirinç 2,5 kg', 18.5, 'gida-yemek-malzemeleri', 'makarna-pirinc-ve-bakliyat', 'pirinc'),
 (2, 25656, 'Reis Jasmine Pirinç 1 kg', 8.5, 'gida-yemek-malzemeleri', 'makarna-pirinc-ve-bakliyat', 'pirinc'),
 (3, 25657, 'Reis Kepekli Diyet Pirinç 1 kg', 8.5, 'gida-yemek-malzemeleri', 'makarna-pirinc-ve-bakliyat', 'pirinc')]

This table has knowledge about items and their categories and price as well.

In [15]:
%%sql
SELECT * FROM BASKET LIMIT 3

 * sqlite:///etrade.db
Done.


[(1, 8938, '2020-07-04 19:39:13.337', '2020-07-04 19:43:55.337', 1),
 (2, 3403, '2021-01-19 19:40:27.557', '2021-01-19 19:46:58.557', 1),
 (3, 7981, '2021-04-16 19:40:27.567', '2021-04-16 19:44:04.567', 1)]

This table has knowledge about basket which indicates that  user add the items to their shopping basket.Status_ meaans the shopping was performed or not.It direcctly relates to user table.

In [17]:
%%sql
SELECT * FROM BASKETDETAIL LIMIT 3

 * sqlite:///etrade.db
Done.


[(1, 1, '2020-07-04 19:40:53.337', 3786, 3, 16.85, 50.55),
 (2, 1, '2020-07-04 19:41:32.337', 6567, 2, 10.9, 21.8),
 (3, 1, '2020-07-04 19:42:24.337', 9896, 2, 3.2, 6.4)]

This table has knowledge about basketdetaiıl which indicates that user add what kind of  items to their shopping basket and it shows how many items is bought and what the total price is.It is directly relates to Basket table.

In [18]:
%%sql
SELECT * FROM ORDER_ LIMIT 3

 * sqlite:///etrade.db
Done.


[(1, 8938, 22282, 1, '2020-07-04 19:43:58.337', 1),
 (2, 3403, 8464, 2, '2021-01-19 19:47:09.557', 1),
 (3, 7981, 19914, 3, '2021-04-16 19:44:13.567', 1)]

This table has knowledge about user's order which indicates that the user tries to buy items which are in their shopping basket.It is directly relates user,address and basket table.

In [19]:
%%sql
SELECT * FROM ORDERDETAIL LIMIT 3

 * sqlite:///etrade.db
Done.


[(1, 1, 1, 3786, 3, 16.85, 50.55),
 (2, 1, 2, 6567, 2, 10.9, 21.8),
 (3, 1, 3, 9896, 2, 3.2, 6.4)]

This table has knowledge about user's order detail which indicates that the user tries to buy items which are in their shopping basket.It is directly relates user,address and basket table.Also,it shows how many items is bought and what the total price is.

In [8]:
%%sql
SELECT * FROM PAYMENT LIMIT 3

 * sqlite:///etrade.db
Done.


[(1, 1, '89.35', 2, '2020-07-04 19:43:58.337', 1, 'OK1737362', None),
 (2, 2, '1560', 5, '2021-01-19 19:47:09.557', 1, 'OK2737693', None),
 (3, 3, '5708.25', 2, '2021-04-16 19:44:13.567', 1, 'OK3737694', None)]

In [9]:
%%sql
SELECT * FROM INVOICE LIMIT 3

 * sqlite:///etrade.db
Done.


[(1, 1, '03062021-193913', '2020-07-08 18:04:30.337', 'CRG03062021-193913', 1),
 (2, 2, '03062021-194027', '2021-01-22 01:10:49.557', 'CRG03062021-194027', 1),
 (3, 3, '03062021-194027', '2021-04-18 14:14:16.567', 'CRG03062021-194027', 1)]

This invoice table has knowledge about user's invoice which includes invoiceno and cargo tracking number.The status show they payment was done or not.

In [207]:
%%sql
SELECT * FROM INVOICEDETAIL LIMIT 3

 * sqlite:///etrade.db
Done.


[(1, 1, 1, 3786, 16.85, 3, 50.55),
 (2, 1, 2, 6567, 10.9, 2, 21.8),
 (3, 1, 3, 9896, 3.2, 2, 6.4)]

This table has knowledge about user's invoice detail which includes items information and the table is related to invoice and orderdetail tables also.Additionally,the table has total price which was paid by user.

## DATA CLEANING&MANIPULATION 

I am gonna analyse table content and reform them if it is necessary.

In [1]:
import pandas as pd

In [2]:
import sqlite3

In [3]:
conn = sqlite3.connect('etrade.db')

In [4]:
cursor=conn.cursor()

In [5]:
### USER TABLE

In [6]:
df_user=pd.read_sql_query("SELECT * FROM user;", conn)

In [7]:
df_user.index

RangeIndex(start=0, stop=10000, step=1)

In [13]:
df_user.isnull().sum()

ID             0
USERNAME_      0
PASSWORD_      0
NAMESURNAME    0
EMAIL          0
GENDER         0
CREATEDDATE    0
BIRTHDATE      0
TELNR1         0
TELNR2         0
dtype: int64

 ### There is no null value in the user table and are 10000 user data in the user table.

In [14]:
df_address=pd.read_sql_query("SELECT * FROM ADDRESS;",conn)

In [15]:
###ADDRESS TABLE

In [16]:
df_address.index

RangeIndex(start=0, stop=24945, step=1)

In [17]:
df_address.isnull().sum()

ID             0
USERID         0
ADDRESSTYPE    0
COUNTRYID      0
CITYID         0
TOWNID         0
DISTRICTID     0
POSTALCODE     0
ADDRESSTEXT    0
dtype: int64

### There is no null value in the address table and are 24945 address data in the address table.

In [18]:
###COUNTRY TABLE

In [19]:

df_country=pd.read_sql_query("SELECT * FROM COUNTRY",conn)

In [20]:
df_country.index

RangeIndex(start=0, stop=1, step=1)

In [21]:
df_country.isnull().sum()

ID         0
COUNTRY    0
dtype: int64

### There is no null value in the country table and is just 1 data in the user table.

In [73]:
###CITY TABLE

In [22]:
df_city=pd.read_sql_query("SELECT * FROM CITY",conn)

In [23]:
df_city=pd.read_sql_query("SELECT * FROM TOWN",conn)

In [24]:
df_city.index

RangeIndex(start=0, stop=957, step=1)

In [25]:
df_city.isnull().sum()

ID        0
CITYID    0
TOWN      0
dtype: int64

### There is no null value in the city table and is just 81 data in the city table which comprise all the cities in Turkey.

In [26]:
###TOWN TABLE

In [27]:
df_town=pd.read_sql_query("SELECT * FROM TOWN",conn)

In [28]:
df_town.index

RangeIndex(start=0, stop=957, step=1)

In [29]:
df_town.isnull().sum()

ID        0
CITYID    0
TOWN      0
dtype: int64

### There is no null value in the town table and are 957 data in the town table which comprise all the towns in Turkey.

In [30]:
###DISTRICT TABLE

In [31]:
df_district=pd.read_sql_query("SELECT * FROM DISTRICT",conn)

In [32]:
df_district.index

RangeIndex(start=0, stop=53145, step=1)

In [33]:
df_district.isnull().sum()

ID          0
TOWNID      0
DISTRICT    0
dtype: int64

### There is no null value in the city table and are 53145  data in the district table which comprise all the districts in Turkey.

In [34]:
###DISTRICT_STREET TABLE

In [35]:
df_district_street=pd.read_sql_query("SELECT * FROM DISTRICT_STREET",conn)

In [36]:
df_district_street.index

RangeIndex(start=0, stop=856682, step=1)

In [37]:
df_district_street.isnull().sum()

ID                  0
CITYID              0
TOWNID              0
DISTRICTID          0
NEIGHBORHOODID      0
CITYNAME            0
TOWNNAME            0
DISTRICTNAME        0
NEIGHBORHOODNAME    0
STREETNAME          0
dtype: int64

### There is no null value in the district_street table and are 856682 data in the district_streeet table which comprise all the district_streeet in Turkey.

In [38]:
###ITEM TABLE

In [39]:
df_item=pd.read_sql_query("SELECT * FROM ITEM",conn)

In [40]:
df_item.index

RangeIndex(start=0, stop=28110, step=1)

df_item.isnull().sum()

### There is no null value in the item table and are 28110 data in the item table.Among these items, the users are going to shop and buy whatever they want.

In [41]:
### BASKET TABLE

In [42]:
df_basket=pd.read_sql_query("SELECT * FROM BASKET",conn)

In [43]:
df_basket.index

RangeIndex(start=0, stop=44919, step=1)

In [44]:
df_city.isnull().sum()

ID        0
CITYID    0
TOWN      0
dtype: int64

### There is no null value in the basket table and  44919 data are in basket data table.

### BASKETDETAIL TABLE

In [45]:
df_basketdetail=pd.read_sql_query("SELECT * FROM BASKETDETAIL",conn)

In [46]:
df_basketdetail.index

RangeIndex(start=0, stop=201315, step=1)

In [47]:
df_basketdetail.isnull().sum()

ID          0
BASKETID    0
DATE_       0
ITEMID      0
AMOUNT      0
PRICE       0
TOTAL       0
dtype: int64

### There is no null value in the city basketdetail table  and are 201315 data in the basketdetail table which comprises all the items the users add to their basket.

In [48]:
###ORDER_ TABLE

In [49]:
df_order=pd.read_sql_query("SELECT * FROM ORDER_",conn)

In [50]:
df_order.index

RangeIndex(start=0, stop=44914, step=1)

In [51]:
df_city.isnull().sum()

ID        0
CITYID    0
TOWN      0
dtype: int64

### There is no null value in the order_ table and are 44914 data in the order_ table which means that the users bought the products were added to their basket.The shopping was performed.

In [52]:
###ORDERDETAIL TABLE

In [53]:
df_orderdetail=pd.read_sql_query("SELECT * FROM ORDERDETAIL",conn)

In [54]:
df_order.index

RangeIndex(start=0, stop=44914, step=1)

In [55]:
df_city.isnull().sum()

ID        0
CITYID    0
TOWN      0
dtype: int64

### There is no null value in the orderdetail table and are 44914 data in the that table which indicates detail of shopping.The order_ table and orderdetail table have same data as can be seen.

In [56]:
###PAYMENT TABLE

In [57]:
df_payment=pd.read_sql_query("SELECT * FROM PAYMENT",conn)

In [58]:
df_payment.index

RangeIndex(start=0, stop=44914, step=1)

In [59]:
df_city.isnull().sum()

ID        0
CITYID    0
TOWN      0
dtype: int64

### There is no null value in the payment table and are 44914 data in the that table which indicates the shopping payment was valid or not.If it is valid, passing to invoice step.

In [60]:
###INVOICE TABLE

In [61]:
df_invoice=pd.read_sql_query("SELECT * FROM INVOICE",conn)

In [62]:
df_invoice.index

RangeIndex(start=0, stop=44914, step=1)

In [63]:
df_city.isnull().sum()

ID        0
CITYID    0
TOWN      0
dtype: int64

In [64]:
df_invoice.isnull().sum()

ID           0
ORDERID      0
INVOICENO    0
DATE_        0
CARGOFNO     0
STATUS_      0
dtype: int64

### There is no null value in the invoice table and are 44914 data in the that table.That table includes information that shows cargo no and creation of the invoice.

In [65]:
###INVOICEDETAIL TABLE

In [66]:
df_invoicedetail=pd.read_sql_query("SELECT * FROM INVOICEDETAIL",conn)

In [67]:
df_invoicedetail.index

RangeIndex(start=0, stop=201301, step=1)

In [68]:
df_invoicedetail.isnull().sum()

ID               0
INVOICEID        0
ORDERDETAILID    0
ITEMID           0
PRICE            0
AMOUNT           0
TOTALPRICE       0
dtype: int64

 There is no null value in the invoicedetail table and are 201301 data in the that table.That table includes information that shows total amount and price of items and related to orderdetail,item and invoice tables directly.

#####  *We have a look over all the tables briefly and try to learn how the tables are big and what kind of information they include. Now, it is time to ask some question and try to derive information which are disguised in the tables.*

#### Question 1: Bring me the users information with their address text and provide information that includes the users address number which are not in the same city.

In [69]:
import pandas as pd

In [70]:
import sqlite3

In [71]:
conn = sqlite3.connect('etrade.db')

In [72]:
cursor=conn.cursor()

In [73]:
df_user=pd.read_sql_query("""SELECT 
U.ID,U.USERNAME_,U.NAMESURNAME,U.EMAIL,U.GENDER,U.BIRTHDATE,U.TELNR1,U.TELNR2,
COUNT(A.ID) TOTALADDRESSNUMBER,COUNT(DISTINCT C.CITY) DIFFERENTCITIESADDRESSNUMBER
FROM user U 
INNER JOIN ADDRESS A ON U.ID=A.USERID
INNER JOIN CITY C ON C.ID=A.CITYID
GROUP BY U.ID,U.USERNAME_,U.NAMESURNAME,U.EMAIL,U.GENDER,U.BIRTHDATE,U.TELNR1,U.TELNR2
HAVING COUNT(A.ID)<>COUNT(DISTINCT C.CITY)
""",conn)

In [74]:
df_user.index

RangeIndex(start=0, stop=671, step=1)

In [75]:
df_user.head()

ID    USERNAME_     NAMESURNAME                      EMAIL GENDER  \
0  34      S_SAPCI    Serdar ŞAPÇI      S_SAPCI@sqlegitim.com      E   
1  52    S_KARIMAN   Sabri KARİMAN    S_KARIMAN@sqlegitim.com      E   
2  55   A_INCIKAPI  Ahmet İNCİKAPI   A_INCIKAPI@sqlegitim.com      E   
3  74  O_EYYVBOGLU  Onur EYYÜBOĞLU  O_EYYVBOGLU@sqlegitim.com      E   
4  90    S_KAMURAN    Sude KAMURAN    S_KAMURAN@sqlegitim.com      K   

    BIRTHDATE        TELNR1        TELNR2  TOTALADDRESSNUMBER  \
0  1961-02-11  (554)1355353  (544)4796414                   3   
1  1952-05-06  (536)9088163  (532)9777077                   2   
2  1991-05-28  (532)2414618  (538)8459085                   4   
3  1945-09-29  (537)9702529  (536)4471784                   4   
4  1967-12-23  (554)8591714  (541)4484973                   4   

   DIFFERENTCITIESADDRESSNUMBER  
0                             2  
1                             1  
2                             3  
3                             3  
4                             3

##### As we know we have 10000 users, just 671 users have addresses that are in different cities.For example Serdar Şapçi has 3 total address, however these 3 addresses exist in 2 cities.

#### Question 2: Bring me the  top users' shopping number,amount and totalprice

In [76]:
df_shopping=pd.read_sql_query("""SELECT 
U.ID,U.USERNAME_,
COUNT(DISTINCT B.ID) NUMBEROFSHOPPING ,
COUNT(B.ID) NUMBEROFITEMS,
SUM(BD.AMOUNT) AMOUNTOFITEMS
,SUM(TOTAL) TOTALPRICE
FROM user U 
INNER JOIN BASKET B ON B.USERID=U.ID
INNER JOIN BASKETDETAIL BD ON BD.BASKETID=B.ID
group by U.ID,U.USERNAME_
ORDER BY SUM(TOTAL) DESC
""",conn)

In [77]:
df_shopping.head()

ID  USERNAME_  NUMBEROFSHOPPING  NUMBEROFITEMS  AMOUNTOFITEMS  TOTALPRICE
0   601  Y_OZSOLAK                 4             27             71    52078.47
1  6579    S_YAVUZ                 5             34             82    51461.78
2  2317    Z_HALAT                 3             19             51    48407.20
3  3963  P_SALAGOZ                 3             24             58    39054.21
4  3215  I_BALIMRE                 5             28             73    38097.99

##### The query shows how many times shopping was performed and how many and how much amount of items are bought and total price of items.

#### Question 3: Bring me the top cities shopping number,amount and totalprice

In [78]:
df_shopping_city=pd.read_sql_query("""SELECT 
C.CITY,
COUNT(DISTINCT B.ID) NUMBEROFSHOPPING ,
COUNT(B.ID) NUMBEROFITEMS,
SUM(BD.AMOUNT) AMOUNTOFITEMS
,SUM(TOTAL) TOTALPRICE
FROM user U 
INNER JOIN ADDRESS A ON U.ID=A.USERID
INNER JOIN CITY C ON C.ID=A.CITYID
INNER JOIN BASKET B ON B.USERID=U.ID
INNER JOIN BASKETDETAIL BD ON BD.BASKETID=B.ID

group by C.CITY
ORDER BY SUM(TOTAL) DESC
""",conn)

In [79]:
df_shopping_city.head()

CITY  NUMBEROFSHOPPING  NUMBEROFITEMS  AMOUNTOFITEMS  TOTALPRICE
0  İSTANBUL              9368          51348         127877  5676069.06
1     İZMİR              5929          31442          78577  3223290.42
2    ANKARA              3944          20754          51735  2266565.54
3     KONYA              3699          19278          48275  2142839.28
4     BURSA              3475          18040          44970  1965123.23

##### The query shows how many times shopping was performed and how many and how much amount of items are bought and total price of items by cities

#### Question 4: Bring me the top brand 

In [80]:
df_shopping_brand=pd.read_sql_query("""SELECT 
I.CATEGORY1,I.CATEGORY2,I.CATEGORY3,
SUM(OD.TOTALPRICE) TOTAL 

FROM ITEM I 
INNER JOIN ORDERDETAIL OD   ON OD.ITEMID=I.ID
INNER JOIN ORDER_ O ON O.ID=OD.ORDERID

group by I.CATEGORY1,I.CATEGORY2,I.CATEGORY3
ORDER BY SUM(OD.TOTALPRICE) DESC
""",conn)

In [81]:
df_shopping_brand.head()

CATEGORY1          CATEGORY2   CATEGORY3       TOTAL
0  elektronik-gida-disi         elektronik        None  5640792.54
1  elektronik-gida-disi         beyaz-esya        None  2734029.10
2  elektronik-gida-disi         elektronik  bilgisayar  1838074.19
3  elektronik-gida-disi  kucuk-ev-aletleri        None  1257945.25
4  elektronik-gida-disi         beyaz-esya   buzdolabi   911494.10

##### The query shows how many times shopping was performed and how many and how much amount of items are bought and total price of items by brands

#### Question 5: Bring me the male and female totalprice who shopped in 2021,January

In [95]:
df_shopping_gender=pd.read_sql_query("""SELECT 
U.GENDER,SUM(BD.TOTAL) FROM user U
INNER JOIN BASKET B ON B.USERID=U.ID
INNER JOIN BASKETDETAIL BD ON BD.BASKETID=B.ID
WHERE B.CREATEDDATE BETWEEN ' 2021-01-01 00:00:00.00' AND '2021-01-31 23:59:59.623'
GROUP BY U.GENDER
""",conn)

TypeError: 'tuple' object is not callable

In [83]:
df_shopping_gender.head()

GENDER  SUM(BD.TOTAL)
0      E     6701712.73
1      K     7663754.06

##### The query shows how many times shopping was performed and how many and how much amount of items are bought and total price of items by gender.

#### Question 6: Bring me top 5 sold  items per city.

In [84]:
df_top5_per_city=pd.read_sql("""
SELECT C.CITY,C1.CATEGORY1, SUM(C1.TOTALPRICE) TOTALPRICE

FROM ORDERDETAIL OD
CROSS JOIN

(SELECT  CATEGORY1,SUM(OD.TOTALPRICE) TOTALPRICE FROM ITEM I
INNER JOIN ORDERDETAIL OD ON OD.ITEMID=I.ID WHERE I.ID=OD.ITEMID GROUP BY CATEGORY1 LIMIT 5 ) C1

INNER JOIN ORDER_ O ON O.ID=OD.ORDERID
INNER JOIN ITEM I ON I.ID=OD.ITEMID
INNER JOIN ADDRESS A ON A.ID=O.ADDRESSID
INNER JOIN CITY C ON C.ID=A.CITYID
GROUP BY C.CITY,C1.CATEGORY1
ORDER BY 1,3 DESC 


""",conn)

In [85]:
df_top5_per_city

CITY               CATEGORY1    TOTALPRICE
0     ADANA    elektronik-gida-disi  1.098249e+11
1     ADANA           kisisel-bakim  1.310147e+10
2     ADANA  gida-yemek-malzemeleri  6.366108e+09
3     ADANA      et-balik-sarkuteri  3.348042e+09
4     ADANA                  icecek  1.873060e+09
..      ...                     ...           ...
400  ŞIRNAK    elektronik-gida-disi  8.656149e+09
401  ŞIRNAK           kisisel-bakim  1.032628e+09
402  ŞIRNAK  gida-yemek-malzemeleri  5.017622e+08
403  ŞIRNAK      et-balik-sarkuteri  2.638850e+08
404  ŞIRNAK                  icecek  1.476303e+08

[405 rows x 3 columns]

##### The query shows TOP 5 products that sold in per city.We have 81 cities that is why we have 405 columns.

#### Question 7: Bring me user's addresses count and show the address which is user's last shopping address

In [86]:
df_user_last_address=pd.read_sql_query("""select U.ID,U.USERNAME_, 
(SELECT COUNT(*) FROM ADDRESS A WHERE A.USERID=U.ID)  ADDRESSCOUNT ,
(SELECT ADDRESSTEXT FROM ADDRESS  WHERE ID IN 
(SELECT   ADDRESSID FROM ORDER_  WHERE USERID=U.ID ORDER BY DATE_ DESC LIMIT 1 )) LASTSHOPPINGADDRESS
 from user U LIMIT 5


""",conn)

In [87]:
df_user_last_address

ID       USERNAME_  ADDRESSCOUNT  \
0   1     N_OZSIMITCI             4   
1   2  A_UNLUMAMULERI             1   
2   3    D_GURBETOGLU             1   
3   4         E_SELIM             2   
4   5         S_VLGEN             2   

                                 LASTSHOPPINGADDRESS  
0  YEN? MAH.FENERL?K SOKAK NO:431 79739 KIZIL�REN...  
1    ATAMER MAH.2581. SOKAK NO:529 96650 KONAK/?ZM?R  
2  FATIH MAH.PLEVNE CADDES? SOKAK NO:563 119342 S...  
3  FAT?H MAH.4583. SOKAK NO:46 94569 ISPARTA MERK...  
4  �UKURKUYU-CAM? MAH.�?�EKL?BEL? SOKAK NO:258 10...

##### The query shows users address and their shopping last address registered in thee database.

#### Question 8: Users have two phone numbers and they use different GSMOPERATORs.In Turkey, if one phone number starts with "50" or "55" it means, it belongs to a spesific GSMOPERATOR.Similarly, if it starts with "54", belongs to another GSMOPERATOR and "53" shows another GSMOPERATOR.Let's count how many users use each GSMOPERATOR. 

In [8]:
df_differentGSMcount=pd.read_sql_query("""SELECT 
SUM(TELNR1_XOPERATORCOUNT+TELNR2_XOPERATORCOUNT) AS XOPERATORCOUNT,
SUM(TELNR1_YOPERATORCOUNT+TELNR2_YOPERATORCOUNT) AS YOPERATORCOUNT,
SUM(TELNR1_ZOPERATORCOUNT+TELNR2_ZOPERATORCOUNT) AS ZOPERATORCOUNT FROM 
(SELECT 
    CASE
        WHEN TELNR1 LIKE '(50%' OR TELNR1 LIKE '(55%' THEN 1 
        ELSE 0
        END AS TELNR1_XOPERATORCOUNT,
    CASE
        WHEN TELNR1 LIKE '(54%' THEN 1 
        ELSE 0
        END AS TELNR1_YOPERATORCOUNT,
    CASE
        WHEN TELNR1 LIKE '(53%' THEN 1 
        ELSE 0
        END AS TELNR1_ZOPERATORCOUNT,
    CASE
        WHEN TELNR2 LIKE '(50%' OR TELNR1 LIKE '(55%' THEN 1 
        ELSE 0
        END AS TELNR2_XOPERATORCOUNT,
    CASE
        WHEN TELNR2 LIKE '(54%' THEN 1
        ELSE 0
        END AS TELNR2_YOPERATORCOUNT,
    CASE
        WHEN TELNR2 LIKE '(53%' THEN 1
        ELSE 0
        END AS TELNR2_ZOPERATORCOUNT
FROM user) T


""",conn)

In [9]:
df_differentGSMcount

XOPERATORCOUNT  YOPERATORCOUNT  ZOPERATORCOUNT
0            4644            5332            9961

##### The query shows, for example X operator has 4644 users,Y operator has 5332 and Z operator has 9961 users in the database

# CONCLUSION


I created a rational database on SQLilte named it Etrade. All the tables are form by me and data for table is finded from  the some web sources.It is better to repeate that all the personel information is simulated.

Here, Just a  few questions were asked and tried to respond it by using just SQL.There is no visualization or creating some report for speficial question.
